core: Jetson test_core fails with some tests from Core_InputOutput

<!--
If you have a question rather than reporting a bug please go to http://answers.opencv.org where you get much faster responses.
If you need further assistance please read [How To Contribute](https://github.com/opencv/opencv/wiki/How_to_contribute).

This is a template helping you to create an issue which can be processed as quickly as possible. This is the bug reporting section for the OpenCV library.
-->

##### System information (version)
<!-- Example
- OpenCV => 3.1
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2015
-->

- OpenCV => recent master (after #9457, including [3.3.1 ](https://github.com/opencv/opencv/releases/tag/3.3.1) )
- Operating System / Platform => Jetson TX2 (ARM 64bit + Ubuntu 16.04)
- Compiler => gcc 5.4.0
 - CUDA => 8.0

##### Detailed description
 - On previously described environment, the ```opencv_test_core``` fails on following tests
   - ```Core_InputOutput.filestorage_keypoints_vec_vec_io```
   - ```Core_InputOutput.FileStorage_DMatch_vector```
   - ```Core_InputOutput.FileStorage_DMatch_vector_vector```
   - ```Core_InputOutput.FileStorage_KeyPoint_vector```
   - ```Core_InputOutput.FileStorage_KeyPoint_vector_vector```
 - For me, it seems that it's comparing with wrong reference
   - This commit ( 164a41b399a2539b066cf741df6f2d911b01ea29 ) seems disabling the new feature when CUDA is specified
```
 ocv_target_compile_definitions(${the_module} PUBLIC OPENCV_TRAITS_ENABLE_DEPRECATED)
```
  - but the actual output seems that the comparison is done to the new output format
```
[ RUN      ] Core_InputOutput.filestorage_keypoints_vec_vec_io
unknown file: Failure
C++ exception with description "/home/nvidia/opencv-fork/modules/core/src/persistence.cpp:4958: error: (-27) Null pointer to reader or destination array in function cvReadRawDataSlice
" thrown in the test body.
[  FAILED  ] Core_InputOutput.filestorage_keypoints_vec_vec_io (0 ms)
[ RUN      ] Core_InputOutput.FileStorage_DMatch
[       OK ] Core_InputOutput.FileStorage_DMatch (1 ms)
[ RUN      ] Core_InputOutput.FileStorage_DMatch_vector
/home/nvidia/opencv-fork/modules/core/test/test_io.cpp:1168: Failure
Expected: fs_read["dv"] >> dv_read doesn't throw an exception.
  Actual: it throws.
[  FAILED  ] Core_InputOutput.FileStorage_DMatch_vector (0 ms)
[ RUN      ] Core_InputOutput.FileStorage_DMatch_vector_vector
/home/nvidia/opencv-fork/modules/core/test/test_io.cpp:1227: Failure
      Expected: fs_result.c_str()
      Which is: "%YAML:1.0\n---\ndvv:\n   - [ [ 1, 2, 3, -1.5000000000000000e+00 ], [ 2, 3, 4,\n       1.5000000000000000e+00 ], [ 3, 2, 1, 5.0000000000000000e-01 ] ]\n   - [ [ 3, 2, 1, 5.0000000000000000e-01 ], [ 1, 2, 3,\n       -1.5000000000000000e+00 ] ]\n"
To be equal to: "%YAML:1.0\n" "---\n" "dvv:\n" "   -\n" "      - [ 1, 2, 3, -1.5000000000000000e+00 ]\n" "      - [ 2, 3, 4, 1.5000000000000000e+00 ]\n" "      - [ 3, 2, 1, 5.0000000000000000e-01 ]\n" "   -\n" "      - [ 3, 2, 1, 5.0000000000000000e-01 ]\n" "      - [ 1, 2, 3, -1.5000000000000000e+00 ]\n"
      Which is: "%YAML:1.0\n---\ndvv:\n   -\n      - [ 1, 2, 3, -1.5000000000000000e+00 ]\n      - [ 2, 3, 4, 1.5000000000000000e+00 ]\n      - [ 3, 2, 1, 5.0000000000000000e-01 ]\n   -\n      - [ 3, 2, 1, 5.0000000000000000e-01 ]\n      - [ 1, 2, 3, -1.5000000000000000e+00 ]\n"
With diff:
@@ -2,6 +2,9 @@
 ---
 dvv:
-   - [ [ 1, 2, 3, -1.5000000000000000e+00 ], [ 2, 3, 4,
-       1.5000000000000000e+00 ], [ 3, 2, 1, 5.0000000000000000e-01 ] ]
-   - [ [ 3, 2, 1, 5.0000000000000000e-01 ], [ 1, 2, 3,
-       -1.5000000000000000e+00 ] ]\n
+   -
+      - [ 1, 2, 3, -1.5000000000000000e+00 ]
+      - [ 2, 3, 4, 1.5000000000000000e+00 ]
+      - [ 3, 2, 1, 5.0000000000000000e-01 ]
+   -
+      - [ 3, 2, 1, 5.0000000000000000e-01 ]
+      - [ 1, 2, 3, -1.5000000000000000e+00 ]\n

/home/nvidia/opencv-fork/modules/core/test/test_io.cpp:1233: Failure
Expected: fs_read["dvv"] >> dvv_read doesn't throw an exception.
  Actual: it throws.
[  FAILED  ] Core_InputOutput.FileStorage_DMatch_vector_vector (1 ms)
[ RUN      ] Core_InputOutput.FileStorage_KeyPoint
[       OK ] Core_InputOutput.FileStorage_KeyPoint (0 ms)
[ RUN      ] Core_InputOutput.FileStorage_KeyPoint_vector
/home/nvidia/opencv-fork/modules/core/test/test_io.cpp:1311: Failure
Expected: fs_read["kv"] >> kv_read doesn't throw an exception.
  Actual: it throws.
[  FAILED  ] Core_InputOutput.FileStorage_KeyPoint_vector (1 ms)
[ RUN      ] Core_InputOutput.FileStorage_KeyPoint_vector_vector
/home/nvidia/opencv-fork/modules/core/test/test_io.cpp:1364: Failure
      Expected: fs_result.c_str()
      Which is: "<?xml version=\"1.0\"?>\n<opencv_storage>\n<kvv>\n  <_><_>\n      1. 2. 16. 0. 100. 1 -1</_>\n    <_>\n      2. 3. 16. 45. 100. 1 -1</_>\n    <_>\n      1. 2. 16. 90. 100. 1 -1</_></_>\n  <_><_>\n      1. 2. 16. 90. 100. 1 -1</_>\n    <_>\n      1. 2. 16. 0. 100. 1 -1</_></_></kvv>\n</opencv_storage>\n"
To be equal to: "<?xml version=\"1.0\"?>\n" "<opencv_storage>\n" "<kvv>\n" "  <_>\n" "    <_>\n" "      1. 2. 16. 0. 100. 1 -1</_>\n" "    <_>\n" "      2. 3. 16. 45. 100. 1 -1</_>\n" "    <_>\n" "      1. 2. 16. 90. 100. 1 -1</_></_>\n" "  <_>\n" "    <_>\n" "      1. 2. 16. 90. 100. 1 -1</_>\n" "    <_>\n" "      1. 2. 16. 0. 100. 1 -1</_></_></kvv>\n" "</opencv_storage>\n"
      Which is: "<?xml version=\"1.0\"?>\n<opencv_storage>\n<kvv>\n  <_>\n    <_>\n      1. 2. 16. 0. 100. 1 -1</_>\n    <_>\n      2. 3. 16. 45. 100. 1 -1</_>\n    <_>\n      1. 2. 16. 90. 100. 1 -1</_></_>\n  <_>\n    <_>\n      1. 2. 16. 90. 100. 1 -1</_>\n    <_>\n      1. 2. 16. 0. 100. 1 -1</_></_></kvv>\n</opencv_storage>\n"
With diff:
@@ -2,5 +2,6 @@
 <opencv_storage>
 <kvv>
-  <_><_>
+  <_>
+    <_>
       1. 2. 16. 0. 100. 1 -1</_>
     <_>
@@ -8,5 +9,6 @@
     <_>
       1. 2. 16. 90. 100. 1 -1</_></_>
-  <_><_>
+  <_>
+    <_>
       1. 2. 16. 90. 100. 1 -1</_>
     <_>

/home/nvidia/opencv-fork/modules/core/test/test_io.cpp:1369: Failure
Expected: fs_read["kvv"] >> kvv_read doesn't throw an exception.
  Actual: it throws.
[  FAILED  ] Core_InputOutput.FileStorage_KeyPoint_vector_vector (1 ms)
```
  - I pasted my entire log [here](https://gist.github.com/tomoaki0705/ade2b1b93c3e0885fc64fa99a575ac58) 
##### Steps to reproduce
  - Build recent OpenCV on Jetson TX2 (same thing happens on Jetson TX1 and TK1)
  - make sure to enable ```WITH_CUDA```
  - run ```opencv_test_core```

<!-- to add code example fence it with triple backticks and optional file extension
    ```.cpp
    // C++ code example
    ```
 or attach as .txt or .zip file
-->
##### notes
   - I should have a look bit more, but since it seems that there is going to be a release of OpenCV 3.4 soon, I though I should raised the notice.

--- a/modules/core/include/opencv2/core/persistence.hpp
+++ b/modules/core/include/opencv2/core/persistence.hpp
@@ -851,7 +851,9 @@ namespace internal
             size_t remaining = it->remaining;
             size_t cn = DataType<_Tp>::channels;
             int _fmt = traits::SafeFmt<_Tp>::fmt;
+            CV_Assert((_fmt >> 8) < 9);
             char fmt[] = { (char)((_fmt >> 8)+'1'), (char)_fmt, '\0' };
+            CV_Assert((remaining % cn) == 0);
             size_t remaining1 = remaining / cn;
             count = count < remaining1 ? count : remaining1;
             vec.resize(count);

--- a/modules/core/src/persistence.cpp
+++ b/modules/core/src/persistence.cpp
@@ -7367,8 +7367,18 @@ void read(const FileNode& node, std::vector<KeyPoint>& keypoints)
     if (first_node.isSeq())
     {
         // modern scheme
+#ifdef OPENCV_TRAITS_ENABLE_DEPRECATED
+        FileNodeIterator it = node.begin();
+        size_t total = (size_t)it.remaining;
+        keypoints.resize(total);
+        for (size_t i = 0; i < total; ++i, ++it)
+        {
+            (*it) >> keypoints[i];
+        }
+#else
         FileNodeIterator it = node.begin();
         it >> keypoints;
+#endif
         return;
     }
     keypoints.clear();
@@ -7394,8 +7404,18 @@ void read(const FileNode& node, std::vector<DMatch>& matches)
     if (first_node.isSeq())
     {
         // modern scheme
+#ifdef OPENCV_TRAITS_ENABLE_DEPRECATED
+        FileNodeIterator it = node.begin();
+        size_t total = (size_t)it.remaining;
+        matches.resize(total);
+        for (size_t i = 0; i < total; ++i, ++it)
+        {
+            (*it) >> matches[i];
+        }
+#else
         FileNodeIterator it = node.begin();
         it >> matches;
+#endif
         return;
     }
     matches.clear();

--- a/modules/core/test/test_io.cpp
+++ b/modules/core/test/test_io.cpp
@@ -1199,6 +1199,7 @@ TEST(Core_InputOutput, FileStorage_DMatch_vector_vector)
 
     EXPECT_NO_THROW(fs << "dvv" << dvv);
     cv::String fs_result = fs.releaseAndGetString();
+#ifndef OPENCV_TRAITS_ENABLE_DEPRECATED
 #if defined _MSC_VER && _MSC_VER <= 1700 /* MSVC 2012 and older */
     EXPECT_STREQ(fs_result.c_str(),
 "%YAML:1.0\n"
@@ -1226,6 +1227,7 @@ TEST(Core_InputOutput, FileStorage_DMatch_vector_vector)
 "      - [ 1, 2, 3, -1.5000000000000000e+00 ]\n"
 );
 #endif
+#endif // OPENCV_TRAITS_ENABLE_DEPRECATED
 
     cv::FileStorage fs_read(fs_result, cv::FileStorage::READ | cv::FileStorage::MEMORY);
 
@@ -1344,6 +1346,7 @@ TEST(Core_InputOutput, FileStorage_KeyPoint_vector_vector)
 
     EXPECT_NO_THROW(fs << "kvv" << kvv);
     cv::String fs_result = fs.releaseAndGetString();
+#ifndef OPENCV_TRAITS_ENABLE_DEPRECATED
     EXPECT_STREQ(fs_result.c_str(),
 "<?xml version=\"1.0\"?>\n"
 "<opencv_storage>\n"
@@ -1362,6 +1365,7 @@ TEST(Core_InputOutput, FileStorage_KeyPoint_vector_vector)
 "      1. 2. 16. 0. 100. 1 -1</_></_></kvv>\n"
 "</opencv_storage>\n"
 );
+#endif //OPENCV_TRAITS_ENABLE_DEPRECATED
 
     cv::FileStorage fs_read(fs_result, cv::FileStorage::READ | cv::FileStorage::MEMORY);
 

